### PR TITLE
[DB-18235][Sizing Calc]: Add all-sharded strategy and choose best out of all-sharded & colo+sharded combined

### PR DESCRIPTION
**Pull request by shaharuk-yb**
### Describe the changes in this pull request
- Add all-sharded strategy code
- add function to choose best strategy out of the two(all-sharded & colo+sharded combined)
- added unit tests
- unit test case fixes
- fixed existing CI tests.

### How Decision Making Works
We initially find out two best sizing recommendations:
- **colocated+sharded**: the current strategy(colocate as much as possible and rest as sharded) 
 - **all-sharded**: put all objects as sharded(new implementation)
  
The `pickBestRecommendationStrategy` function is responsible for selecting the optimal sizing strategy between two approaches: **colocated+sharded** and **all-sharded** strategies. It uses a following decision-making process that prioritizes efficiency and resource optimization.

#### Decision Logic Flow

1. **Failure Handling**: First checks if either strategy has failure reasoning (e.g., cannot support requirements). If both fail, returns an error message. If only one fails, selects the working strategy.

2. **Resultant Cores Comparison**: When strategies have different VCPUs per instance, calculates `resultant_cores = VCPUsPerInstance × NumNodes` and:
   - Selects the strategy with fewer resultant cores (more efficient)
   - If equal, prefers the strategy with higher VCPUs per instance (better performance)

3. **Node Count Fallback**: When VCPUs per instance are the same, compares by number of nodes:
   - Selects the strategy requiring fewer nodes
   - If equal, prefers all-sharded strategy (simpler architecture)

4. **Detailed Reporting**: Provides comprehensive reasoning including object counts (tables and indexes) for both selected and non-selected strategies.

### Test Scenarios

| Test Function | Scenario | Description |
|---------------|----------|-------------|
| `TestPickBestRecommendationStrategy_VCPULogic_Rec1FewerResultantCores` | VCPU Logic - Rec1 Fewer Resultant Cores | Verifies selection of colocated+sharded strategy when it has fewer resultant cores (4×3=12 vs 8×2=16) |
| `TestPickBestRecommendationStrategy_VCPULogic_EqualResultantCores_SelectHigherVCPUs` | VCPU Logic - Equal Resultant Cores | Tests selection of higher VCPUs when resultant cores are equal (6×2=12 vs 4×3=12) |
| `TestPickBestRecommendationStrategy_VCPULogic_Rec2FewerResultantCores` | VCPU Logic - Rec2 Fewer Resultant Cores | Verifies selection of all-sharded strategy when it has fewer resultant cores (8×4=32 vs 16×3=48) |
| `TestPickBestRecommendationStrategy_VCPULogic_Rec1FewerVCPUsButMoreResultantCores` | VCPU Logic - Rec1 Fewer VCPUs But More Resultant Cores | Tests selection based on resultant cores despite lower VCPUs (4×7=28 vs 8×3=24) |
| `TestPickBestRecommendationStrategy_VCPULogic_Rec2FewerVCPUsButMoreResultantCores` | VCPU Logic - Rec2 Fewer VCPUs But More Resultant Cores | Tests selection based on resultant cores despite lower VCPUs (16×3=48 vs 8×7=56) |
| `TestPickBestRecommendationStrategy_SameVCPUs_UseNodeComparison_Rec1FewerNodes` | Same VCPUs - Rec1 Fewer Nodes | Tests node comparison fallback when VCPUs are equal (3 nodes vs 5 nodes) |
| `TestPickBestRecommendationStrategy_SameVCPUs_UseNodeComparison_Rec2FewerNodes` | Same VCPUs - Rec2 Fewer Nodes | Tests node comparison fallback when VCPUs are equal (4 nodes vs 6 nodes) |
| `TestPickBestRecommendationStrategy_SameVCPUsAndNodes_PreferAllSharded` | Same VCPUs and Nodes | Tests preference for all-sharded when both strategies are equivalent |
| `TestPickBestRecommendationStrategy_WithIndexes_VerifyObjectCounts` | With Indexes - Verify Object Counts | Validates accurate counting of tables and indexes in reasoning output |
| `TestPickBestRecommendationStrategy_EdgeCase_EqualResultantCores` | Edge Case - Equal Resultant Cores | Tests edge case with zero cores needed and equal resultant cores |
| `TestPickBestRecommendationStrategy_EdgeCase_LargeValues` | Edge Case - Large Values | Tests function with large-scale values (64 VCPUs, 50+ nodes) |
| `TestPickBestRecommendationStrategy_ComplexScenario_WithMultipleIndexes` | Complex Scenario - Multiple Indexes | Tests complex scenario with multiple tables and indexes |
| `TestPickBestRecommendationStrategy_ReasoningFormat_ResultantCoresComparison` | Reasoning Format - Resultant Cores | Validates reasoning message format for resultant cores comparison |
| `TestPickBestRecommendationStrategy_SameVCPUs_FallbackToNodeComparison` | Same VCPUs - Fallback to Node Comparison | Tests fallback to node comparison when VCPUs are identical |
| `TestPickBestRecommendationStrategy_BothHaveFailureReasoning` | Both Have Failure Reasoning | Tests error handling when both strategies fail |
| `TestPickBestRecommendationStrategy_Rec1HasFailureRec2DoesNot` | Rec1 Has Failure, Rec2 Does Not | Tests selection of working strategy when one fails |
| `TestPickBestRecommendationStrategy_Rec2HasFailureRec1DoesNot` | Rec2 Has Failure, Rec1 Does Not | Tests selection of working strategy when one fails |
| `TestPickBestRecommendationStrategy_BothHaveEmptyFailureReasoning` | Both Have Empty Failure Reasoning | Tests normal comparison logic when no failures exist |
### How was this pull request tested?
- manual
- CI jobs
- Unit tests

### Does your PR have changes that can cause upgrade issues? 
| Component        | Breaking changes? |
| :----------------------------------------------: | :-----------: |
| MetaDB                                                    |  No |
| Name registry json                                        |  No |
| Data File Descriptor Json                                 |  No |
| Export Snapshot Status Json                               |  No |
| Import Data State                                         |  No |
| Export Status Json                                        |  No |
| Data .sql files of tables                                 |  No |
| Export and import data queue                              |  No |
| Schema Dump                                               |  No |
| AssessmentDB                                              |  No |
| Sizing DB                                                 |  No |
| Migration Assessment Report Json                          |  No |
| Callhome Json                                             |  No |
| YugabyteD Tables                                          |  No |
| TargetDB Metadata Tables                                  |  No |

### CODE REVIEW

File: yb-voyager/src/migassessment/sizing.go [hunk 1]
```diff
@@ -214,21 +214,35 @@ func SizingAssessment(targetDbVersion *ybversion.YBVersion, sourceDBType string)
 		return fmt.Errorf("error fetching the sharded throughput: %w", err)
 	}
 
-	sizingRecommendationPerCore := createSizingRecommendationStructure(colocatedLimits)
+	sizingRecommendationPerCoreColocatedAndShardedCombined := createRecommendationStructureColocatedAndShardedCombined(colocatedLimits)
```
Comment (makalaaneesh) [inline yb-voyager/src/migassessment/sizing.go:217]: could we group all the strategy operations together? It would be easier to follow, that way! 

i.e. first we have all the steps for the `colocatedAndShardedCombined` strategy up until we calculate the best recommendation for this strategy. And then, we have all the steps for the `allSharded` strategy. 
Finally, we choose the best between the two.

File: yb-voyager/src/migassessment/sizing.go [hunk 2]
```diff
@@ -214,21 +214,35 @@ func SizingAssessment(targetDbVersion *ybversion.YBVersion, sourceDBType string)
...
+
+	sizingRecommendationPerCoreAllSharded = findNumNodesNeededBasedOnTabletsRequired(sourceIndexMetadata, shardedLimits, sizingRecommendationPerCoreAllSharded)
+
+	finalSizingRecommendationColoShardedCombined := pickBestRecommendation(sizingRecommendationPerCoreColocatedAndShardedCombined)
+
+	finalSizingRecommendationAllSharded := pickBestRecommendation(sizingRecommendationPerCoreAllSharded)
+
+	finalSizingRecommendation, _ := pickBestRecommendationStrategy(finalSizingRecommendationColoShardedCombined, finalSizingRecommendationAllSharded, sourceIndexMetadata)
+
+	if finalSizingRecommendationColoShardedCombined.FailureReasoning != "" {
```
Comment (makalaaneesh) [inline yb-voyager/src/migassessment/sizing.go:245]: this should be finalSizingRecommendation, right?

File: yb-voyager/src/migassessment/sizing.go [hunk 3]
```diff
@@ -214,21 +214,35 @@ func SizingAssessment(targetDbVersion *ybversion.YBVersion, sourceDBType string)
...
+
+	sizingRecommendationPerCoreColocatedAndShardedCombined = findNumNodesNeededBasedOnTabletsRequired(sourceIndexMetadata, shardedLimits, sizingRecommendationPerCoreColocatedAndShardedCombined)
+
+	sizingRecommendationPerCoreAllSharded = findNumNodesNeededBasedOnTabletsRequired(sourceIndexMetadata, shardedLimits, sizingRecommendationPerCoreAllSharded)
+
+	finalSizingRecommendationColoShardedCombined := pickBestRecommendation(sizingRecommendationPerCoreColocatedAndShardedCombined)
+
+	finalSizingRecommendationAllSharded := pickBestRecommendation(sizingRecommendationPerCoreAllSharded)
+
+	finalSizingRecommendation, _ := pickBestRecommendationStrategy(finalSizingRecommendationColoShardedCombined, finalSizingRecommendationAllSharded, sourceIndexMetadata)
```
Comment (makalaaneesh) [inline yb-voyager/src/migassessment/sizing.go:243]: we don't seem to be using the second return argument? shall we remove it then?

File: yb-voyager/src/migassessment/sizing.go [hunk 4]
```diff
@@ -375,6 +389,125 @@ func pickBestRecommendation(recommendations map[int]IntermediateRecommendation)
...
+  - recAllSharded: Second IntermediateRecommendation to compare
+  - sourceIndexMetadata: A slice of SourceDBMetadata structs representing source indexes
+
+Returns:
+  - The best IntermediateRecommendation based on the selection logic described above
+  - A string explaining the reasoning behind the choice, including comparison details and detailed object counts
+*/
+func pickBestRecommendationStrategy(recCombined, recAllSharded IntermediateRecommendation, sourceIndexMetadata []SourceDBMetadata) (IntermediateRecommendation, string) {
+	// Handle failure reasoning logic first
+	rec1HasFailure := recCombined.FailureReasoning != ""
```
Comment (makalaaneesh) [inline yb-voyager/src/migassessment/sizing.go:413]: nit: shall we not use rec1/2 and instead stick to `recCombined`/ `recSharded`. It becomes a bit easier to follow

File: yb-voyager/src/migassessment/sizing.go [hunk 5]
```diff
@@ -375,6 +389,125 @@ func pickBestRecommendation(recommendations map[int]IntermediateRecommendation)
...
+			selectedRec = recAllSharded
+			notSelectedRec = recCombined
+			reasoning = fmt.Sprintf("\nSelected all-sharded strategy with %d VCPUs per instance (%.1f resultant cores) over colocated+sharded strategy with %d VCPUs per instance (%.1f resultant cores). ",
+				recAllSharded.VCPUsPerInstance, rec2ResultantCores, recCombined.VCPUsPerInstance, rec1ResultantCores)
+			vcpuConditionMet = true
+		}
+	}
+
+	// If VCPUsPerInstance condition was not met, use original node comparison logic
+	if !vcpuConditionMet {
```
Comment (makalaaneesh) [inline yb-voyager/src/migassessment/sizing.go:474]: do we really need this condition variable? seems like a simple if-else would work, no?

File: yb-voyager/src/migassessment/sizing.go [hunk 6]
```diff
@@ -375,6 +389,125 @@ func pickBestRecommendation(recommendations map[int]IntermediateRecommendation)
...
+		} else {
+			// Equal cores - prefer all-sharded (recAllSharded)
+			selectedRec = recAllSharded
+			notSelectedRec = recCombined
+			reasoning = fmt.Sprintf("\nSelected all-sharded strategy (same %.0f nodes as colocated+sharded strategy, preferring all-sharded). ",
+				recAllSharded.NumNodes)
+		}
+	}
+
+	// Get detailed object counts for non-selected recommendations
```
Comment (makalaaneesh) [inline yb-voyager/src/migassessment/sizing.go:495]: if the "reasoning" info was just for debugging, can we convert all these to logs, instead of keeping them in the "reasoning" variable? 

or at least let's log the reasoning variable ?

File: yb-voyager/src/migassessment/sizing_test.go [hunk 7]
```diff
@@ -941,6 +941,495 @@ func TestFindNumNodesNeededBasedOnTabletsRequired_NeedMoreNodes(t *testing.T) {
...
+		VCPUsPerInstance: 8,
+		CoresNeeded:      16,
+		NumNodes:         2,
+		ColocatedTables:  []SourceDBMetadata{{ObjectName: "table3"}},
+		ShardedTables:    []SourceDBMetadata{{ObjectName: "table4"}},
+		FailureReasoning: "Exceeds table limits",
+	}
+	var sourceIndexMetadata []SourceDBMetadata
+
+	_, reasoning := pickBestRecommendationStrategy(rec1, rec2, sourceIndexMetadata)
```
Comment (makalaaneesh) [inline yb-voyager/src/migassessment/sizing_test.go:1349]: nit: assert that one of them is returned (whichever you're arbitrarily picking)

File: yb-voyager/src/migassessment/sizing.go (hunk 2 above)
Comment (shaharuk-yb) [reply yb-voyager/src/migassessment/sizing.go]: yes, corrected it now.

File: yb-voyager/src/migassessment/sizing.go (hunk 4 above)
Comment (shaharuk-yb) [reply yb-voyager/src/migassessment/sizing.go]: done, updated it to rec1 , rec2

File: yb-voyager/src/migassessment/sizing_test.go (hunk 7 above)
Comment (shaharuk-yb) [reply yb-voyager/src/migassessment/sizing_test.go]: updated the test case.

File: yb-voyager/src/migassessment/sizing.go (hunk 4 above)
Comment (makalaaneesh) [reply yb-voyager/src/migassessment/sizing.go]: (nit) I think you mis-read my comment. I was suggesting we stick to `recCombined/recSharded` because 1/2 can be confusing.

**Review by makalaaneesh — APPROVED** [review]
Logic LGTM! 
pls also get a sign-off from shubham before merging wrt the reports. 
Also, I believe some test cases are still failing ( because of the missing colocated clause)
